### PR TITLE
Developer experience issues for FML

### PR DESCRIPTION
**Pull request by jhugman**
Fixes [EXP-2250](https://mozilla-hub.atlassian.net/browse/EXP-2250)
Fixes [EXP-2251](https://mozilla-hub.atlassian.net/browse/EXP-2251)
Fixes [EXP-2252](https://mozilla-hub.atlassian.net/browse/EXP-2252)
Fixes [EXP-2253](https://mozilla-hub.atlassian.net/browse/EXP-2253)

This PR fixes several papercuts and developer experience issues:

 * property level defaults are type checked for objects now.
 * enum maps and objects defined at the feature property level can overwrite feature level property defaults.
 * enum maps don't need to be complete within the feature's `defaults` blocks.
 * improves error messages to add the path of where validation errors was detected.
 * Allow Strings to be used as defaults in Text values for Android resources.

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/support/nimbus-fml/src/error.rs [hunk 1]
```diff
@@ -25,8 +25,8 @@ pub enum FMLError {
     #[allow(dead_code)]
     #[error("Internal error: {0}")]
     InternalError(&'static str),
-    #[error("Validation Error: {0}")]
-    ValidationError(String),
+    #[error("Validation Error at {0}: {1}")]
+    ValidationError(String, String),
```
Comment (tarikeshaq) [inline components/support/nimbus-fml/src/error.rs:29]: 🤩 I was wrestling with this code a while ago to see if I can pull line numbers, but that sounded like a **very** big pain. I'm still hoping that one day we can do that (probably means not using serde_yaml 😞)

But regardless thank you for this!!

File: components/support/nimbus-fml/src/parser.rs [hunk 2]
```diff
@@ -36,30 +39,46 @@ pub(crate) struct FieldBody {
...
 #[derive(Debug, Deserialize, Serialize, Clone)]
+#[serde(deny_unknown_fields)]
 pub(crate) struct ObjectBody {
     description: String,
     failable: Option<bool>,
     fields: HashMap<String, FieldBody>,
 }
 
 #[derive(Debug, Deserialize, Serialize, Clone, Default)]
+#[serde(deny_unknown_fields)]
```
Comment (tarikeshaq) [inline components/support/nimbus-fml/src/parser.rs:50]: Looking at https://serde.rs/field-attrs.html#flatten, it's saying flatten shouldn't be used with `deny_unknown_fields`, not sure what the implication is but we seem to have good test coverage so 🤷

File: components/support/nimbus-fml/src/parser.rs [hunk 3]
```diff
@@ -256,109 +278,188 @@ fn collect_channel_defaults(
...
-/// # Returns
-/// Returns a transformed [`FeatureDef`] with its defaults merged
-pub fn merge_feature_defaults(
-    feature_def: FeatureDef,
-    channel: &str,
-    supported_channels: &[String],
-) -> Result<FeatureDef, FMLError> {
-    if !supported_channels.iter().any(|c| c == channel) {
-        return Err(FMLError::InvalidChannelError(channel.into()));
+struct DefaultsMerger<'object> {
```
Comment (tarikeshaq) [inline components/support/nimbus-fml/src/parser.rs:281]: Yeah the merging was and is getting complex 😕 I feel like we're missing a generic struct or trait to help us out - for a future refactor something like: 

```rs
trait Merger {
  type: T1;
  type: T2;
  fn merge(higher_priority: T1, lower_priority: T2) -> serde_json::Value;
}

struct DefaultMerger<T1, T2>;

impl Merger for DefaultMerger<PropDefaults, ObjectDefaults> {
..
}

impl Merger for DefaultMerger<FeatureDefaults, PropDefaults> {
...
}

```

might help reshape this a little

**Review by tarikeshaq — APPROVED** [review]
Looks good to me, I dropped a few thoughts as I was going but nothing blocking!

File: components/support/nimbus-fml/src/parser.rs (hunk 3 above)
Comment (jhugman) [reply components/support/nimbus-fml/src/parser.rs:281]: Yeah, there's some refactoring to be done here.

We have a `Defaults` trait in `nimbus`, which would be helpful.

I'm relatively sure that there's material work needed to pull out the feature's property defaults and the default blocks into data structures that can be merged, and then back into the feature's property defaults. The merging once in that common data structure is relatively straightfoward.

File: components/support/nimbus-fml/fixtures/fe/dx_improvements.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        type: String
+        default: A string
+      an-object:
+        description: A variable of type TestObject
+        type: TestObject
+        default:
+          name: Feature default
+          optional-int: 5
+          enum-map:
+            alice: 11
```
[the remark this answers is not in the document]
Comment (jhugman) [reply components/support/nimbus-fml/fixtures/fe/dx_improvements.yaml:19]: Yeah, I think this was possible all along.

File: components/support/nimbus-fml/src/backends/kotlin/gen_structs/bundled.rs [hunk 5]
```diff
@@ -73,6 +77,10 @@ impl CodeType for TextCodeType {
     }
 }
 
+fn is_resource_id(string: &str) -> bool {
+    !string.contains("://") && !string.contains(' ')
```
[the remark this answers is not in the document]
Comment (jhugman) [reply components/support/nimbus-fml/src/backends/kotlin/gen_structs/bundled.rs:81]: added a comment and some tests.

File: components/support/nimbus-fml/src/backends/kotlin/gen_structs/common.rs [hunk 6]
```diff
@@ -20,8 +20,13 @@ pub fn enum_variant_name(nm: &dyn Display) -> String {
 }
 
 /// Surrounds a property name with quotes. It is assumed that property names do not need escaping.
-pub fn quoted(prop: &dyn Display) -> String {
-    format!("\"{}\"", prop)
+pub fn quoted(string: &dyn Display) -> String {
+    let string = string.to_string();
+    if string.contains('"') || string.contains('\n') {
+        format!(r#""""{}""""#, string)
```
[the remark this answers is not in the document]
Comment (jhugman) [reply components/support/nimbus-fml/src/backends/kotlin/gen_structs/common.rs:26]: Added a comment and a test.

**Pull request merged**